cpp/src/text/bpe/byte_pair_encoding.cuh

```diff
@@ -30,85 +31,165 @@
+  __device__ bool operator()(cudf::size_type lhs, merge_pair_type const& rhs) const noexcept
   {
-    return d_strings.element<cudf::string_view>(lhs) == rhs;
+    lhs *= 2;
+    auto const left  = d_strings.element<cudf::string_view>(lhs);
+    auto const right = d_strings.element<cudf::string_view>(lhs + 1);
+    return (left == rhs.first) && (right == rhs.second);
   }
 };
 
-using hash_table_allocator_type = rmm::mr::stream_allocator_adaptor<default_allocator<char>>;
-
-using probe_scheme = cuco::experimental::linear_probing<1, bpe_hasher>;
+using bpe_probe_scheme = cuco::experimental::linear_probing<1, bpe_hasher>;
 
 using merge_pairs_map_type = cuco::experimental::static_map<cudf::size_type,
                                                             cudf::size_type,
                                                             cuco::experimental::extent<std::size_t>,
                                                             cuda::thread_scope_device,
                                                             bpe_equal,
-                                                            probe_scheme,
+                                                            bpe_probe_scheme,
                                                             hash_table_allocator_type>;
 
+/**
+ * @brief Hasher function used for building and using the cuco static-map
+ *
+ * This takes advantage of heterogeneous lookup feature in cuco static-map which
+ * allows inserting with one type (index) and looking up with a different type (merge_pair_type).
+ *
+ * Each component of the merge-pairs (left and right) are stored individually in the map.
+ */
+struct mp_hasher {
+  cudf::column_device_view const d_strings;
+  string_hasher_type hasher{};
+  // used by insert
+  __device__ hash_value_type operator()(cudf::size_type index) const
+  {
+    auto const d_str = d_strings.element<cudf::string_view>(index);
+    return hasher(d_str);
+  }
+  // used by find
+  __device__ hash_value_type operator()(cudf::string_view const& d_str) const
+  {
+    return hasher(d_str);
+  }
+};
+
+/**
+ * @brief Equal function used for building and using the cuco static-map
+ *
+ * This takes advantage of heterogeneous lookup feature in cuco static-map which
+ * allows inserting with one type (index) and looking up with a different type (string).
+ */
+struct mp_equal {
+  cudf::column_device_view const d_strings;
+  // used by insert
+  __device__ bool operator()(cudf::size_type lhs, cudf::size_type rhs) const noexcept
+  {
+    auto const left  = d_strings.element<cudf::string_view>(lhs);
+    auto const right = d_strings.element<cudf::string_view>(rhs);
+    return left == right;
+  }
+  // used by find
+  __device__ bool operator()(cudf::size_type lhs, cudf::string_view const& rhs) const noexcept
+  {
+    auto const left = d_strings.element<cudf::string_view>(lhs);
+    return left == rhs;
+  }
+};
+
+using mp_probe_scheme = cuco::experimental::linear_probing<1, mp_hasher>;
```

Comment:
nit: this is the same as `mp_probe_scheme`. maybe unify them?

Reply:
Do you mean `bpe_probe_scheme`? 
One uses `mp_hasher` and the other uses `bpe_hasher` which are very different.
Maybe I'm missing something.

Reply:
Ah, you are right, I thought that they were both linear probing with CG size = 1 and didn't pay attention to the hasher part.